// node-broadlink/dist/alarm.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S1C = void 0;
var device_1 = __importDefault(require("./device"));
var S1C = /** @class */ (function (_super) {
    __extends(S1C, _super);
    function S1C() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'S1C';
        _this.S1C_SENSORS_TYPES = {
            0x31: 'Door Sensor',
            0x91: 'Key Fob',
            0x21: 'Motion Sensor',
        };
        return _this;
    }
    S1C.prototype.getSensorsStatus = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var packet = Buffer.alloc(16);
            packet[0] = 0x06;
            _this.sendPacketAndDecrypt(packet)
                .then(function (response) {
                var payload = _this.decrypt(response);
                var count = payload[0x4];
                var sensors = payload.subarray(0x6);
                var sensorsA = __spreadArray([], __read(Array(Math.floor(sensors.length / 83)).keys()), false).map(function (i) {
                    return sensors.subarray(i * 83, (i + 1) * 83);
                });
                resolve({
                    count: count,
                    sensors: sensorsA.reduce(function (acc, sensor) {
                        var serial = sensor.subarray(26, 30).toString('hex');
                        if (serial !== '00000000') {
                            acc.push({
                                status: sensor[0],
                                name: response.subarray(4, 26).toString().trim(),
                                type: _this.S1C_SENSORS_TYPES[sensor[3]] || 'Unknown',
                                order: sensor[1],
                                serial: serial,
                            });
                        }
                        return acc;
                    }, []),
                });
            })
                .catch(reject);
        });
    };
    return S1C;
}(device_1.default));
exports.S1C = S1C;
