// node-broadlink/dist/climate.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hysen = void 0;
var device_1 = __importDefault(require("./device"));
var Hysen = /** @class */ (function (_super) {
    __extends(Hysen, _super);
    function Hysen() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'Hysen heating controller';
        return _this;
    }
    Hysen.prototype.calculateCrc16 = function (inputData) {
        var crc16Tab = __spreadArray([], __read(Array(256).keys()), false).map(function (i) {
            return __spreadArray([], __read(Array(8)), false).reduce(function (crc) { return (crc & 0x0001 ? (crc >> 1) ^ 0xa001 : crc >> 1); }, i);
        });
        return __spreadArray([], __read(inputData), false).reduce(function (crcValue, c) { return (crcValue >> 8) ^ crc16Tab[(crcValue ^ c) & 0x00ff]; }, 0xffff);
    };
    Hysen.toBytes2 = function (value) {
        return Buffer.of(value & 0xff, (value >> 8) & 0xff);
    };
    Hysen.prototype.sendRequest = function (inputPayload) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.sendPacket(Buffer.concat([
                Hysen.toBytes2(inputPayload.length + 2),
                inputPayload,
                Hysen.toBytes2(_this.calculateCrc16(inputPayload)),
            ]))
                .then(function (response) {
                var responsePayload = _this.decrypt(response);
                var responsePayloadLength = responsePayload[0];
                try {
                    if (responsePayloadLength + 2 > responsePayload.length) {
                        throw new Error('hysenResponseError first byte of response is not length');
                    }
                    else {
                        var nom_crc = response.subarray(0, 2).readInt8();
                        var real_crc = _this.calculateCrc16(responsePayload.subarray(2, responsePayloadLength));
                        if (nom_crc !== real_crc) {
                            resolve(responsePayload.subarray(2, responsePayloadLength));
                        }
                        else {
                            throw new Error('hysenResponseError CRC check on response failed');
                        }
                    }
                }
                catch (err) {
                    reject(err);
                }
            })
                .catch(reject);
        });
    };
    Hysen.prototype.getTemp = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.sendRequest(Buffer.of(0x01, 0x03, 0x00, 0x00, 0x00, 0x08))
                .then(function (payload) {
                resolve(payload[0x05] / 2.0);
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Hysen.prototype.getExternalTemp = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.sendRequest(Buffer.of(0x01, 0x03, 0x00, 0x00, 0x00, 0x08))
                .then(function (payload) {
                resolve(payload[18] / 2.0);
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Hysen.prototype.getFullStatus = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.sendRequest(Buffer.of(0x01, 0x03, 0x00, 0x00, 0x00, 0x16))
                .then(function (payload) {
                var week = __spreadArray([], __read(Array(8).keys()), false).map(function (i) {
                    return {
                        startHour: payload[2 * i + 23],
                        startMinute: payload[2 * i + 24],
                        temp: payload[i + 39] / 2.0,
                    };
                });
                var roomTempAdj = ((payload[13] << 8) + payload[14]) / 2.0;
                resolve({
                    remoteLock: payload[3] & 1,
                    power: payload[4] & 1,
                    active: (payload[4] >> 4) & 1,
                    tempManual: (payload[4] >> 6) & 1,
                    roomTemp: (payload[5] & 255) / 2.0,
                    thermostatTemp: (payload[6] & 255) / 2.0,
                    autoMode: payload[7] & 15,
                    loopMode: (payload[7] >> 4) & 15,
                    sensor: payload[8],
                    osv: payload[9],
                    dif: payload[10],
                    svh: payload[11],
                    svl: payload[12],
                    roomTempAdj: roomTempAdj > 32767 ? 32767 - roomTempAdj : roomTempAdj,
                    fre: payload[15],
                    poweron: payload[16],
                    unknown: payload[17],
                    externalTemp: (payload[18] & 255) / 2.0,
                    hour: payload[19],
                    min: payload[20],
                    sec: payload[21],
                    dayofweek: payload[22],
                    weekDay: week.slice(0, 6),
                    weekEnd: week.slice(6),
                });
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Hysen.prototype.setMode = function (autoMode, loopMode, sensor) {
        if (sensor === void 0) { sensor = 0; }
        return this.sendRequest(Buffer.of(0x01, 0x06, 0x00, 0x02, ((loopMode + 1) << 4) + autoMode, sensor)).then();
    };
    Hysen.prototype.setAdvanced = function (loopMode, sensor, osv, dif, svh, svl, adj, fre, powerOn) {
        return this.sendRequest(Buffer.of(0x01, 0x10, 0x00, 0x02, 0x00, 0x05, 0x0a, loopMode, sensor, osv, dif, svh, svl, ((adj * 10) >> 8) & 0xff, (adj * 10) & 0xff, fre, powerOn)).then();
    };
    Hysen.prototype.setTemp = function (temp) {
        return this.sendRequest(Buffer.of(0x01, 0x06, 0x00, 0x01, 0x00, temp * 2)).then();
    };
    Hysen.prototype.setPower = function (power, remoteLock) {
        if (power === void 0) { power = 1; }
        if (remoteLock === void 0) { remoteLock = 0; }
        return this.sendRequest(Buffer.of(0x01, 0x06, 0x00, 0x00, remoteLock, power)).then();
    };
    Hysen.prototype.setTime = function (hour, minute, second, day) {
        return this.sendRequest(Buffer.of(0x01, 0x10, 0x00, 0x08, 0x00, 0x02, 0x04, hour, minute, second, day)).then();
    };
    Hysen.prototype.setSchedule = function (weekday, weekend) {
        return this.sendRequest(Buffer.from(__spreadArray(__spreadArray(__spreadArray(__spreadArray([
            0x01,
            0x10,
            0x00,
            0x0a,
            0x00,
            0x0c,
            0x18
        ], __read(__spreadArray([], __read(Array(6).keys()), false).map(function (i) { return [weekday[i].startHour, weekday[i].startMinute]; }).flat()), false), __read(__spreadArray([], __read(Array(2).keys()), false).map(function (i) { return [weekend[i].startHour, weekend[i].startMinute]; }).flat()), false), __read(__spreadArray([], __read(Array(6).keys()), false).map(function (i) { return weekday[i].temp * 2; })), false), __read(__spreadArray([], __read(Array(2).keys()), false).map(function (i) { return weekend[i].temp * 2; })), false))).then();
    };
    return Hysen;
}(device_1.default));
exports.Hysen = Hysen;
