// node-broadlink/dist/cover.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dooya = void 0;
var device_1 = __importDefault(require("./device"));
var Dooya = /** @class */ (function (_super) {
    __extends(Dooya, _super);
    function Dooya() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'Dooya DT360E';
        return _this;
    }
    Dooya.prototype.open = function () {
        return this.send(0x01, 0x00);
    };
    Dooya.prototype.close = function () {
        return this.send(0x02, 0x00);
    };
    Dooya.prototype.stop = function () {
        return this.send(0x03, 0x00);
    };
    Dooya.prototype.getPercentage = function () {
        return this.send(0x06, 0x5d);
    };
    Dooya.prototype.setPercentageAndWait = function (newPercentage) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.getPercentage()
                .then(function (current) {
                if (current > newPercentage) {
                    _this.close()
                        .then(function () {
                        var interval = setInterval(function () {
                            _this.getPercentage()
                                .then(function (current) {
                                if (current === undefined || current <= newPercentage) {
                                    clearInterval(interval);
                                    resolve(_this.stop());
                                }
                            })
                                .catch(function (err) {
                                clearInterval(interval);
                                reject(err);
                            });
                        }, 200);
                    })
                        .catch(function (err) {
                        reject(err);
                    });
                }
                else if (current < newPercentage) {
                    _this.open()
                        .then(function () {
                        var interval = setInterval(function () {
                            _this.getPercentage()
                                .then(function (current) {
                                if (current === undefined || current >= newPercentage) {
                                    clearInterval(interval);
                                    resolve(_this.stop());
                                }
                            })
                                .catch(function (err) {
                                clearInterval(interval);
                                reject(err);
                            });
                        }, 200);
                    })
                        .catch(function (err) {
                        reject(err);
                    });
                }
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Dooya.prototype.send = function (magic1, magic2) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var packet = Buffer.alloc(16);
            packet[0] = 0x09;
            packet[2] = 0xbb;
            packet[3] = magic1;
            packet[4] = magic2;
            packet[9] = 0xfa;
            packet[10] = 0x44;
            _this.sendPacket(packet)
                .then(function (response) {
                resolve(_this.decrypt(response)[4]);
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    return Dooya;
}(device_1.default));
exports.Dooya = Dooya;
