// node-broadlink/dist/device.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var dgram_1 = __importDefault(require("dgram"));
var crypto_1 = __importDefault(require("crypto"));
var Device = /** @class */ (function () {
    function Device(host, mac, deviceType, model, manufacturer, name, isLocked) {
        if (deviceType === void 0) { deviceType = 0x272a; }
        if (model === void 0) { model = ''; }
        if (manufacturer === void 0) { manufacturer = ''; }
        if (name === void 0) { name = ''; }
        if (isLocked === void 0) { isLocked = false; }
        this.host = host;
        this.mac = mac;
        this.deviceType = deviceType;
        this.model = model;
        this.manufacturer = manufacturer;
        this.name = name;
        this.isLocked = isLocked;
        this.id = [0, 0, 0, 0];
        this.key = [0x09, 0x76, 0x28, 0x34, 0x3f, 0xe9, 0x9e, 0x23, 0x76, 0x5c, 0x15, 0x13, 0xac, 0xcf, 0x8b, 0x02];
        this.count = Math.floor(Math.random() * 0xffff);
        this.iv = Buffer.from([
            0x56, 0x2e, 0x17, 0x99, 0x6d, 0x09, 0x3d, 0x28, 0xdd, 0xb3, 0xba, 0x69, 0x5a, 0x2e, 0x6f, 0x58,
        ]);
        this.TYPE = 'Unknown';
        this.socket = dgram_1.default.createSocket('udp4');
        this.socket.bind();
    }
    Device.prototype.encrypt = function (payload) {
        var cipher = crypto_1.default.createCipheriv('aes-128-cbc', Buffer.from(this.key), this.iv);
        return cipher.update(payload);
    };
    Device.prototype.decrypt = function (response) {
        var decipher = crypto_1.default.createDecipheriv('aes-128-cbc', Buffer.from(this.key), this.iv);
        decipher.setAutoPadding(false);
        return Buffer.concat([decipher.update(response.subarray(0x38)), decipher.final()]);
    };
    Device.prototype.auth = function () {
        var _this = this;
        var payload = Buffer.alloc(0x50);
        payload.fill(0x31, 0x04, 0x14);
        payload[0x1e] = 0x01;
        payload[0x2d] = 0x01;
        payload.fill('Test 1', 0x30, 0x36);
        return new Promise(function (resolve, reject) {
            _this.sendPacket(payload, 0x65)
                .then(function (response) {
                var responsePayload = _this.decrypt(response);
                var key = responsePayload.subarray(0x04, 0x14);
                if (key.length > 0 && key.length % 16 === 0) {
                    _this.id = __spreadArray([], __read(responsePayload.subarray(0x00, 0x04)), false);
                    _this.key = __spreadArray([], __read(key), false);
                    resolve(_this);
                }
                else {
                    reject(new Error('auth failed'));
                }
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Device.prototype.sendPacket = function (buffer, command) {
        var _this = this;
        if (command === void 0) { command = 0x6a; }
        return new Promise(function (resolve, reject) {
            var _a, _b, _c;
            _this.count = (_this.count + 1) & 0xffff;
            var packet = Buffer.alloc(0x38);
            _a = __read([0x5a, 0xa5, 0xaa, 0x55, 0x5a, 0xa5, 0xaa, 0x55], 8), packet[0x00] = _a[0], packet[0x01] = _a[1], packet[0x02] = _a[2], packet[0x03] = _a[3], packet[0x04] = _a[4], packet[0x05] = _a[5], packet[0x06] = _a[6], packet[0x07] = _a[7];
            packet[0x24] = _this.deviceType & 0xff;
            packet[0x25] = (_this.deviceType >> 8) & 0xff;
            packet[0x26] = command;
            packet[0x28] = _this.count & 0xff;
            packet[0x29] = (_this.count >> 8) & 0xff;
            _b = __read(Buffer.from(_this.mac).reverse(), 6), packet[0x2a] = _b[0], packet[0x2b] = _b[1], packet[0x2c] = _b[2], packet[0x2d] = _b[3], packet[0x2e] = _b[4], packet[0x2f] = _b[5];
            _c = __read(_this.id, 4), packet[0x30] = _c[0], packet[0x31] = _c[1], packet[0x32] = _c[2], packet[0x33] = _c[3];
            var payload = Buffer.concat([buffer, Buffer.alloc((((16 - buffer.length) % 16) + 16) % 16)]);
            var checksum = payload.reduce(function (acc, b) { return acc + b; }, 0xbeaf) & 0xffff;
            packet[0x34] = checksum & 0xff;
            packet[0x35] = (checksum >> 8) & 0xff;
            packet = Buffer.concat([packet, _this.encrypt(payload)]);
            checksum = packet.reduce(function (acc, b) { return acc + b; }, 0xbeaf) & 0xffff;
            packet[0x20] = checksum & 0xff;
            packet[0x21] = (checksum >> 8) & 0xff;
            _this.socket.once('message', function (response) {
                var error = response.subarray(0x22, 0x24);
                var errorCode = error[0] | (error[1] << 8);
                if (!errorCode) {
                    resolve(response);
                }
                else {
                    reject(new Error("".concat(errorCode)));
                }
            });
            _this.socket.send(packet, 0, packet.length, _this.host.port, _this.host.address);
        });
    };
    Device.prototype.sendPacketAndDecrypt = function (buffer) {
        var _this = this;
        return this.sendPacket(buffer).then(function (response) { return _this.decrypt(response); });
    };
    return Device;
}());
exports.default = Device;
