// node-broadlink/dist/hub.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3 = void 0;
var device_1 = __importDefault(require("./device"));
var struct_1 = __importDefault(require("./struct"));
var S3 = /** @class */ (function (_super) {
    __extends(S3, _super);
    function S3() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'S3';
        _this.step = 5;
        return _this;
    }
    S3.prototype.range = function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, 0];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, this.step];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, 8];
                case 3:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    };
    S3.prototype.getSubDevices = function () {
        return this.getSubDevicesRecursive([], this.range());
    };
    S3.prototype.getSubDevicesRecursive = function (acc, rangeIterator) {
        var _this = this;
        var _a = rangeIterator.next(), index = _a.value, done = _a.done;
        return this.sendPacketAndDecrypt(this.encode(14, {
            count: this.step,
            index: index,
        }))
            .then(function (data) { return _this.decode(data); })
            .then(function (state) {
            acc.push.apply(acc, __spreadArray([], __read(state.list), false));
            if (acc.length === state.total || done) {
                return acc;
            }
            return _this.getSubDevicesRecursive(acc, rangeIterator);
        });
    };
    S3.prototype.getState = function (did) {
        var _this = this;
        return this.sendPacketAndDecrypt(this.encode(1, { did: did })).then(function (data) {
            return _this.decode(data);
        });
    };
    S3.prototype.setState = function (state) {
        var _this = this;
        return this.sendPacketAndDecrypt(this.encode(2, state)).then(function (response) { return _this.decode(response); });
    };
    S3.prototype.encode = function (flag, state) {
        var data = JSON.stringify(this.mapToNumber(state));
        var packet = Buffer.concat([
            (0, struct_1.default)('<HHHBBI').pack_into(Buffer.alloc(12), 0, 0xa5a5, 0x5a5a, 0, flag, 0x0b, data.length),
            Buffer.from(data),
        ]);
        var checksum = packet.reduce(function (acc, b) { return (acc + b) & 0xffff; }, 0xbeaf);
        packet[0x04] = checksum & 0xff;
        packet[0x05] = (checksum >> 8) & 0xff;
        return packet;
    };
    S3.prototype.decode = function (response) {
        var _a = __read((0, struct_1.default)('<I').unpack_from(response, 0x08), 1), js_len = _a[0];
        return this.mapToBoolean(JSON.parse(response.subarray(0x0c, 0x0c + js_len).toString()));
    };
    S3.prototype.mapToNumber = function (state) {
        return __assign(__assign({}, state), { pwr1: state.pwr1 !== undefined ? Number(state.pwr1) : undefined, pwr2: state.pwr2 !== undefined ? Number(state.pwr2) : undefined, pwr3: state.pwr3 !== undefined ? Number(state.pwr3) : undefined });
    };
    S3.prototype.mapToBoolean = function (state) {
        return __assign(__assign({}, state), { pwr1: state.pwr1 !== undefined ? Boolean(state.pwr1) : undefined, pwr2: state.pwr2 !== undefined ? Boolean(state.pwr2) : undefined, pwr3: state.pwr3 !== undefined ? Boolean(state.pwr3) : undefined });
    };
    return S3;
}(device_1.default));
exports.S3 = S3;
