// node-broadlink/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.discover = exports.genDevice = exports.setup = exports.getNetworkInterfaces = void 0;
var dgram_1 = __importDefault(require("dgram"));
var os_1 = __importDefault(require("os"));
var device_1 = __importDefault(require("./device"));
var remote_1 = require("./remote");
var alarm_1 = require("./alarm");
var climate_1 = require("./climate");
var cover_1 = require("./cover");
var light_1 = require("./light");
var sensor_1 = require("./sensor");
var switch_1 = require("./switch");
var hub_1 = require("./hub");
// Export all TypeScript classes, to avoid internal (node-broadlink/dist/...) import for Liberty users */
__exportStar(require("./alarm"), exports);
__exportStar(require("./climate"), exports);
__exportStar(require("./cover"), exports);
__exportStar(require("./device"), exports);
__exportStar(require("./light"), exports);
__exportStar(require("./remote"), exports);
__exportStar(require("./sensor"), exports);
__exportStar(require("./switch"), exports);
var SUPPORTED_TYPES = {
    0x0000: [switch_1.Sp1, 'SP1', 'Broadlink'],
    0x2717: [switch_1.Sp2, 'NEO', 'Ankuoo'],
    0x2719: [switch_1.Sp2, 'SP2-compatible', 'Honeywell'],
    0x271a: [switch_1.Sp2, 'SP2-compatible', 'Honeywell'],
    0x2720: [switch_1.Sp2, 'SP mini', 'Broadlink'],
    0x2728: [switch_1.Sp2, 'SP2-compatible', 'URANT'],
    0x273e: [switch_1.Sp2, 'SP mini', 'Broadlink'],
    0x7530: [switch_1.Sp2, 'SP2', 'Broadlink (OEM)'],
    0x7539: [switch_1.Sp2, 'SP2-IL', 'Broadlink (OEM)'],
    0x753e: [switch_1.Sp2, 'SP mini 3', 'Broadlink'],
    0x7540: [switch_1.Sp2, 'MP2', 'Broadlink'],
    0x7544: [switch_1.Sp2, 'SP2-CL', 'Broadlink'],
    0x7546: [switch_1.Sp2, 'SP2-UK/BR/IN', 'Broadlink (OEM)'],
    0x7547: [switch_1.Sp2, 'SC1', 'Broadlink'],
    0x7918: [switch_1.Sp2, 'SP2', 'Broadlink (OEM)'],
    0x7919: [switch_1.Sp2, 'SP2-compatible', 'Honeywell'],
    0x791a: [switch_1.Sp2, 'SP2-compatible', 'Honeywell'],
    0x7d0d: [switch_1.Sp2, 'SP mini 3', 'Broadlink (OEM)'],
    0x2711: [switch_1.Sp2s, 'SP2', 'Broadlink'],
    0x2716: [switch_1.Sp2s, 'NEO PRO', 'Ankuoo'],
    0x271d: [switch_1.Sp2s, 'Ego', 'Efergy'],
    0x2736: [switch_1.Sp2s, 'SP mini+', 'Broadlink'],
    0x2733: [switch_1.Sp3, 'SP3', 'Broadlink'],
    0x7d00: [switch_1.Sp3, 'SP3-EU', 'Broadlink (OEM)'],
    0x9479: [switch_1.Sp3s, 'SP3S-US', 'Broadlink'],
    0x947a: [switch_1.Sp3s, 'SP3S-EU', 'Broadlink'],
    0x7568: [switch_1.Sp4, 'SP4L-CN', 'Broadlink'],
    0x756c: [switch_1.Sp4, 'SP4M', 'Broadlink'],
    0x756f: [switch_1.Sp4, 'MCB1', 'Broadlink'],
    0x7579: [switch_1.Sp4, 'SP4L-EU', 'Broadlink'],
    0x757b: [switch_1.Sp4, 'SP4L-AU', 'Broadlink'],
    0x7583: [switch_1.Sp4, 'SP mini 3', 'Broadlink'],
    0x7587: [switch_1.Sp4, 'SP4L-UK', 'Broadlink'],
    0x7d11: [switch_1.Sp4, 'SP mini 3', 'Broadlink'],
    0xa569: [switch_1.Sp4, 'SP4L-UK', 'Broadlink'],
    0xa56a: [switch_1.Sp4, 'MCB1', 'Broadlink'],
    0xa56b: [switch_1.Sp4, 'SCB1E', 'Broadlink'],
    0xa56c: [switch_1.Sp4, 'SP4L-EU', 'Broadlink'],
    0xa589: [switch_1.Sp4, 'SP4L-UK', 'Broadlink'],
    0xa5d3: [switch_1.Sp4, 'SP4L-EU', 'Broadlink'],
    0x5115: [switch_1.Sp4b, 'SCB1E', 'Broadlink'],
    0x51e2: [switch_1.Sp4b, 'AHC/U-01', 'BG Electrical'],
    0x6111: [switch_1.Sp4b, 'MCB1', 'Broadlink'],
    0x6113: [switch_1.Sp4b, 'SCB1E', 'Broadlink'],
    0x618b: [switch_1.Sp4b, 'SP4L-EU', 'Broadlink'],
    0x6489: [switch_1.Sp4b, 'SP4L-AU', 'Broadlink'],
    0x648b: [switch_1.Sp4b, 'SP4M-US', 'Broadlink'],
    0x648c: [switch_1.Sp4b, 'SP4L-US', 'Broadlink'],
    0x6494: [switch_1.Sp4b, 'SCB2', 'Broadlink'],
    0x2737: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x278f: [remote_1.Rmmini, 'RM mini', 'Broadlink'],
    0x27c2: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x27c7: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x27cc: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x27cd: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x27d0: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x27d1: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x27d3: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x27dc: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x27de: [remote_1.Rmmini, 'RM mini 3', 'Broadlink'],
    0x2712: [remote_1.Rmpro, 'RM pro/pro+', 'Broadlink'],
    0x272a: [remote_1.Rmpro, 'RM pro', 'Broadlink'],
    0x273d: [remote_1.Rmpro, 'RM pro', 'Broadlink'],
    0x277c: [remote_1.Rmpro, 'RM home', 'Broadlink'],
    0x2783: [remote_1.Rmpro, 'RM home', 'Broadlink'],
    0x2787: [remote_1.Rmpro, 'RM pro', 'Broadlink'],
    0x278b: [remote_1.Rmpro, 'RM plus', 'Broadlink'],
    0x2797: [remote_1.Rmpro, 'RM pro+', 'Broadlink'],
    0x279d: [remote_1.Rmpro, 'RM pro+', 'Broadlink'],
    0x27a1: [remote_1.Rmpro, 'RM plus', 'Broadlink'],
    0x27a6: [remote_1.Rmpro, 'RM plus', 'Broadlink'],
    0x27a9: [remote_1.Rmpro, 'RM pro+', 'Broadlink'],
    0x27c3: [remote_1.Rmpro, 'RM pro+', 'Broadlink'],
    0x5f36: [remote_1.Rmminib, 'RM mini 3', 'Broadlink'],
    0x6507: [remote_1.Rmminib, 'RM mini 3', 'Broadlink'],
    0x6508: [remote_1.Rmminib, 'RM mini 3', 'Broadlink'],
    0x51da: [remote_1.Rm4mini, 'RM4 mini', 'Broadlink'],
    0x5209: [remote_1.Rm4mini, 'RM4 TV mate', 'Broadlink'],
    0x520c: [remote_1.Rm4mini, 'RM4 mini', 'Broadlink'],
    0x520d: [remote_1.Rm4mini, 'RM4C mini', 'Broadlink'],
    0x5211: [remote_1.Rm4mini, 'RM4C mate', 'Broadlink'],
    0x5212: [remote_1.Rm4mini, 'RM4 TV mate', 'Broadlink'],
    0x5216: [remote_1.Rm4mini, 'RM4 mini', 'Broadlink'],
    0x521c: [remote_1.Rm4mini, 'RM4 mini', 'Broadlink'],
    0x6070: [remote_1.Rm4mini, 'RM4C mini', 'Broadlink'],
    0x610e: [remote_1.Rm4mini, 'RM4 mini', 'Broadlink'],
    0x610f: [remote_1.Rm4mini, 'RM4C mini', 'Broadlink'],
    0x62bc: [remote_1.Rm4mini, 'RM4 mini', 'Broadlink'],
    0x62be: [remote_1.Rm4mini, 'RM4C mini', 'Broadlink'],
    0x6364: [remote_1.Rm4mini, 'RM4S', 'Broadlink'],
    0x648d: [remote_1.Rm4mini, 'RM4 mini', 'Broadlink'],
    0x6539: [remote_1.Rm4mini, 'RM4C mini', 'Broadlink'],
    0x653a: [remote_1.Rm4mini, 'RM4 mini', 'Broadlink'],
    0x520b: [remote_1.Rm4pro, 'RM4 pro', 'Broadlink'],
    0x5213: [remote_1.Rm4pro, 'RM4 pro', 'Broadlink'],
    0x5218: [remote_1.Rm4pro, 'RM4C pro', 'Broadlink'],
    0x6026: [remote_1.Rm4pro, 'RM4 pro', 'Broadlink'],
    0x6184: [remote_1.Rm4pro, 'RM4C pro', 'Broadlink'],
    0x61a2: [remote_1.Rm4pro, 'RM4 pro', 'Broadlink'],
    0x649b: [remote_1.Rm4pro, 'RM4 pro', 'Broadlink'],
    0x653c: [remote_1.Rm4pro, 'RM4 pro', 'Broadlink'],
    0x2714: [sensor_1.A1, 'e-Sensor', 'Broadlink'],
    0x4eb5: [switch_1.Mp1, 'MP1-1K4S', 'Broadlink'],
    0x4ef7: [switch_1.Mp1, 'MP1-1K4S', 'Broadlink (OEM)'],
    0x4f1b: [switch_1.Mp1, 'MP1-1K3S2U', 'Broadlink (OEM)'],
    0x4f65: [switch_1.Mp1, 'MP1-1K3S2U', 'Broadlink'],
    0x5043: [light_1.Lb1, 'SB800TD', 'Broadlink (OEM)'],
    0x504e: [light_1.Lb1, 'LB1', 'Broadlink'],
    0x606e: [light_1.Lb1, 'SB500TD', 'Broadlink (OEM)'],
    0x60c7: [light_1.Lb1, 'LB1', 'Broadlink'],
    0x60c8: [light_1.Lb1, 'LB1', 'Broadlink'],
    0x6112: [light_1.Lb1, 'LB1', 'Broadlink'],
    0x644b: [light_1.Lb1, 'LB1', 'Broadlink'],
    0x644c: [light_1.Lb1, 'LB27 R1', 'Broadlink'],
    0x644e: [light_1.Lb1, 'LB26 R1', 'Broadlink'],
    0xa4f4: [light_1.Lb2, 'LB27 R1', 'Broadlink'],
    0xa5f7: [light_1.Lb2, 'LB27 R1', 'Broadlink'],
    0x2722: [alarm_1.S1C, 'S2KIT', 'Broadlink'],
    0xa59c: [hub_1.S3, 'S3', 'Broadlink'],
    0xa64d: [hub_1.S3, 'S3', 'Broadlink'],
    0x4ead: [climate_1.Hysen, 'HY02/HY03', 'Hysen'],
    0x4e4d: [cover_1.Dooya, 'DT360E-45/20', 'Dooya'],
    0x51e3: [switch_1.Bg1, 'BG800/BG900', 'BG Electrical'],
};
function getNetworkInterfaces(interfaces) {
    return interfaces && (!Array.isArray(interfaces) || interfaces.length > 0)
        ? ((Array.isArray(interfaces) && interfaces) || [interfaces]).flat().map(function (arg) {
            return ({
                address: arg.address || arg,
                broadcastAddress: arg.broadcastAddress || '255.255.255.255',
            });
        })
        : Object.values(os_1.default.networkInterfaces())
            .flat()
            .filter(function (networkInterface) { return networkInterface && networkInterface.family === 'IPv4' && !networkInterface.internal; })
            .map(function (networkInterface) {
            var address = networkInterface.address.split('.');
            return {
                address: networkInterface.address,
                broadcastAddress: networkInterface.netmask
                    .split('.')
                    .map(function (byte, index) { return (byte === '255' ? address[index] : '255'); })
                    .join('.'),
            };
        });
}
exports.getNetworkInterfaces = getNetworkInterfaces;
function setup(ssid, password, securityMode, networkInterfaces) {
    return new Promise(function (resolve, reject) {
        var payload = Buffer.concat([
            Buffer.alloc(68),
            Buffer.of.apply(Buffer, __spreadArray([], __read(ssid.split('').map(function (letter) { return letter.charCodeAt(0); })), false)),
            Buffer.alloc(100 - 68 - ssid.length),
            Buffer.of.apply(Buffer, __spreadArray([], __read(password.split('').map(function (letter) { return letter.charCodeAt(0); })), false)),
            Buffer.alloc(0x88 - 100 - password.length),
        ]);
        payload[0x26] = 0x14;
        payload[0x84] = ssid.length;
        payload[0x85] = password.length;
        payload[0x86] = securityMode;
        var checksum = payload.reduce(function (acc, b) { return acc + b; }, 0xbeaf) & 0xffff;
        payload[0x20] = checksum & 0xff;
        payload[0x21] = (checksum >> 8) & 0xff;
        Promise.all(getNetworkInterfaces(networkInterfaces).map(function (networkInterface) {
            return new Promise(function (resolveSocketBound, rejectSocketBound) {
                var socket = dgram_1.default.createSocket('udp4');
                socket.once('listening', function () {
                    socket.setBroadcast(true);
                    socket.send(payload, 0, payload.length, 80, networkInterface.broadcastAddress);
                });
                socket.on('error', function (err) {
                    socket.close();
                    rejectSocketBound(err);
                });
                socket.bind({ address: networkInterface.address }, function () {
                    socket.close();
                    resolveSocketBound(networkInterface);
                });
            });
        }))
            .then(function (interfaces) {
            resolve({
                ssid: ssid,
                password: password,
                securityMode: "".concat(securityMode, " (").concat(['none', 'WEP', 'WPA1', 'WPA2', 'WPA1/2'][securityMode], ")"),
                interfaces: interfaces,
            });
        })
            .catch(reject);
    });
}
exports.setup = setup;
function genDevice(deviceType, host, mac, name, isLocked) {
    if (deviceType in SUPPORTED_TYPES) {
        var _a = __read(SUPPORTED_TYPES[deviceType], 3), DeviceClazz = _a[0], model = _a[1], manufacturer = _a[2];
        return new DeviceClazz(host, mac, deviceType, model, manufacturer, name, isLocked);
    }
    return new device_1.default(host, mac, deviceType, undefined, undefined, name, isLocked);
}
exports.genDevice = genDevice;
function discover(timeout, interfaces, discoverIpPort) {
    if (timeout === void 0) { timeout = 500; }
    if (discoverIpPort === void 0) { discoverIpPort = 80; }
    return new Promise(function (resolve) {
        var devices = [];
        var sockets = getNetworkInterfaces(interfaces).map(function (networkInterface) {
            var cs = dgram_1.default.createSocket('udp4');
            cs.once('listening', function () {
                cs.setBroadcast(true);
                var address = networkInterface.address.split('.');
                var port = cs.address().port;
                var now = new Date();
                var timezone = now.getTimezoneOffset() / -3600;
                var year = now.getFullYear() - 1900;
                var packet = Buffer.alloc(0x30);
                if (timezone < 0) {
                    packet[0x08] = 0xff + timezone - 1;
                    packet[0x09] = 0xff;
                    packet[0x0a] = 0xff;
                    packet[0x0b] = 0xff;
                }
                else {
                    packet[0x08] = timezone;
                    packet[0x09] = 0;
                    packet[0x0a] = 0;
                    packet[0x0b] = 0;
                }
                packet[0x0c] = year & 0xff;
                packet[0x0d] = (year >> 8) & 0xff;
                packet[0x0e] = now.getMinutes();
                packet[0x0f] = now.getHours();
                packet[0x10] = ~~year % 100;
                packet[0x11] = now.getDay();
                packet[0x12] = now.getDay();
                packet[0x13] = now.getMonth();
                packet[0x18] = ~~address[0];
                packet[0x19] = ~~address[1];
                packet[0x1a] = ~~address[2];
                packet[0x1b] = ~~address[3];
                packet[0x1c] = port & 0xff;
                packet[0x1d] = (port >> 8) & 0xff;
                packet[0x26] = 6;
                var checksum = packet.reduce(function (acc, b) { return acc + b; }, 0xbeaf) & 0xffff;
                packet[0x20] = checksum & 0xff;
                packet[0x21] = (checksum >> 8) & 0xff;
                cs.send(packet, 0, packet.length, discoverIpPort, networkInterface.broadcastAddress);
            });
            cs.on('message', function (msg, rinfo) {
                var deviceType = msg[0x34] | (msg[0x35] << 8);
                var mac = __spreadArray([], __read(msg.subarray(0x3a, 0x40)), false).reverse();
                if (!devices.some(function (device) { return device.mac.toString() === mac.toString(); })) {
                    var nameSlice = msg.slice(0x40, 0x7e);
                    var name = nameSlice.slice(0, nameSlice.indexOf(0x00)).toString('utf8');
                    var isLocked = !!msg[0x7f];
                    devices.push(genDevice(deviceType, rinfo, mac, name, isLocked));
                }
            });
            cs.bind({ address: networkInterface.address });
            return cs;
        });
        setTimeout(function () {
            sockets.forEach(function (socket) { return socket.close(); });
            resolve(devices);
        }, ~~timeout);
    });
}
exports.discover = discover;
