// node-broadlink/dist/light.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lb2 = exports.Lb1 = void 0;
var device_1 = __importDefault(require("./device"));
var struct_1 = __importDefault(require("./struct"));
var Lb = /** @class */ (function (_super) {
    __extends(Lb, _super);
    function Lb() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Lb.prototype.getState = function () {
        var _this = this;
        return this.sendPacketAndDecrypt(this.encode(1, {})).then(function (data) {
            return _this.decode(data);
        });
    };
    Lb.prototype.mapToNumber = function (state) {
        return {
            pwr: state.pwr !== undefined ? Number(state.pwr) : undefined,
            red: state.red,
            blue: state.blue,
            green: state.green,
            brightness: state.brightness,
            colortemp: state.colortemp,
            hue: state.hue,
            saturation: state.saturation,
            transitionduration: state.transitionduration,
            maxworktime: state.maxworktime,
            bulb_colormode: state.bulb_colormode,
            bulb_scenes: state.bulb_scenes,
            bulb_scene: state.bulb_scene,
        };
    };
    Lb.prototype.mapToBoolean = function (state) {
        return {
            pwr: state.pwr !== undefined ? Boolean(state.pwr) : undefined,
            red: state.red,
            blue: state.blue,
            green: state.green,
            brightness: state.brightness,
            colortemp: state.colortemp,
            hue: state.hue,
            saturation: state.saturation,
            transitionduration: state.transitionduration,
            maxworktime: state.maxworktime,
            bulb_colormode: state.bulb_colormode,
            bulb_scenes: state.bulb_scenes,
            bulb_scene: state.bulb_scene,
        };
    };
    Lb.prototype.setState = function (state) {
        var _this = this;
        return this.sendPacket(this.encode(2, state)).then(function (response) { return _this.decode(response); });
    };
    return Lb;
}(device_1.default));
var Lb1 = /** @class */ (function (_super) {
    __extends(Lb1, _super);
    function Lb1() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'SmartBulb';
        return _this;
    }
    Lb1.prototype.mapToNumber = function (state) {
        return __assign(__assign({}, _super.prototype.mapToNumber.call(this, state)), { bulb_sceneidx: state.bulb_sceneidx });
    };
    Lb1.prototype.mapToBoolean = function (state) {
        return __assign(__assign({}, _super.prototype.mapToBoolean.call(this, state)), { bulb_sceneidx: state.bulb_sceneidx });
    };
    Lb1.prototype.encode = function (flag, state) {
        var data = JSON.stringify(this.mapToNumber(state));
        var p_len = 12 + data.length;
        var packet = Buffer.concat([
            (0, struct_1.default)('<HHHHBBI').pack_into(Buffer.alloc(14), 0, p_len, 0xa5a5, 0x5a5a, 0, flag, 0x0b, data.length),
            Buffer.from(data),
        ]);
        var checksum = packet.subarray(0x02).reduce(function (acc, b) { return (acc + b) & 0xffff; }, 0xbeaf);
        packet[0x06] = checksum & 0xff;
        packet[0x07] = (checksum >> 8) & 0xff;
        return packet;
    };
    Lb1.prototype.decode = function (response) {
        var _a = __read((0, struct_1.default)('<I').unpack_from(response, 0xa), 1), js_len = _a[0];
        return this.mapToBoolean(JSON.parse(response.subarray(0xe, 0xe + js_len).toString()));
    };
    return Lb1;
}(Lb));
exports.Lb1 = Lb1;
var Lb2 = /** @class */ (function (_super) {
    __extends(Lb2, _super);
    function Lb2() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'LB2';
        return _this;
    }
    Lb2.prototype.encode = function (flag, state) {
        var data = JSON.stringify(this.mapToNumber(state));
        var packet = Buffer.concat([
            (0, struct_1.default)('<HHHBBI').pack_into(Buffer.alloc(12), 0, 0xa5a5, 0x5a5a, 0, flag, 0x0b, data.length),
            Buffer.from(data),
        ]);
        var checksum = packet.reduce(function (acc, b) { return (acc + b) & 0xffff; }, 0xbeaf);
        packet[0x04] = checksum & 0xff;
        packet[0x05] = (checksum >> 8) & 0xff;
        return packet;
    };
    Lb2.prototype.decode = function (response) {
        var _a = __read((0, struct_1.default)('<I').unpack_from(response, 0x08), 1), js_len = _a[0];
        return this.mapToBoolean(JSON.parse(response.subarray(0x0c, 0x0c + js_len).toString()));
    };
    return Lb2;
}(Lb));
exports.Lb2 = Lb2;
