// node-broadlink/dist/remote.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rm4pro = exports.Rm4mini = exports.Rmminib = exports.Rmpro = exports.Rmmini = void 0;
var device_1 = __importDefault(require("./device"));
var struct_1 = __importDefault(require("./struct"));
function hexStringToPulseArray(data) {
    var _a;
    // Take each 2 chars and convert them to a number
    return (_a = data.match(/[\da-f]{2}/gi)) === null || _a === void 0 ? void 0 : _a.map(function (hex) { return parseInt(hex, 16); });
}
var Rmmini = /** @class */ (function (_super) {
    __extends(Rmmini, _super);
    function Rmmini() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'RMMINI';
        return _this;
    }
    Rmmini.prototype.send = function (command, data) {
        if (data === void 0) { data = []; }
        return this.sendPacketAndDecrypt(Buffer.concat([(0, struct_1.default)('<I').pack(command), Buffer.from(data)])).then(function (payload) {
            return payload.subarray(0x4);
        });
    };
    Rmmini.prototype.checkData = function () {
        return this.send(0x4);
    };
    Rmmini.prototype.sendData = function (data) {
        var pulseArray;
        if (typeof data === 'string') {
            pulseArray = hexStringToPulseArray(data);
        }
        else {
            pulseArray = data;
        }
        return this.send(0x2, pulseArray).then();
    };
    Rmmini.prototype.enterLearning = function () {
        return this.send(0x3).then();
    };
    Rmmini.prototype.cancelLearning = function () {
        return this.send(0x1e).then();
    };
    return Rmmini;
}(device_1.default));
exports.Rmmini = Rmmini;
var Rmpro = /** @class */ (function (_super) {
    __extends(Rmpro, _super);
    function Rmpro() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'RMPRO';
        return _this;
    }
    Rmpro.prototype.sweepFrequency = function () {
        return this.send(0x19).then();
    };
    Rmpro.prototype.cancelSweepFrequency = function () {
        return this.send(0x1e).then();
    };
    Rmpro.prototype.checkFrequency = function () {
        return this.send(0x1a).then(function (response) { return response[0] == 1; });
    };
    Rmpro.prototype.findRfPacket = function () {
        return this.send(0x1b).then();
    };
    Rmpro.prototype.checkTemperature = function () {
        return this.readSensor().then(function (_a) {
            var temperature = _a.temperature;
            return temperature;
        });
    };
    Rmpro.prototype.readSensor = function () {
        return this.send(0x1).then(function (resp) {
            var _a = __read((0, struct_1.default)('<bb').unpack(resp.subarray(0, 0x2)), 2), a = _a[0], b = _a[1];
            return {
                temperature: a + b / 10.0,
            };
        });
    };
    return Rmpro;
}(Rmmini));
exports.Rmpro = Rmpro;
var Rmminib = /** @class */ (function (_super) {
    __extends(Rmminib, _super);
    function Rmminib() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'RMMINIB';
        return _this;
    }
    Rmminib.prototype.send = function (command, data) {
        if (data === void 0) { data = []; }
        return this.sendPacketAndDecrypt(Buffer.concat([(0, struct_1.default)('<HI').pack(data.length + 4, command), Buffer.from(data)])).then(function (payload) {
            return payload.subarray(0x6, (0, struct_1.default)('<H').unpack(payload.subarray(0, 0x2))[0] + 2);
        });
    };
    return Rmminib;
}(Rmmini));
exports.Rmminib = Rmminib;
var Rm4mini = /** @class */ (function (_super) {
    __extends(Rm4mini, _super);
    function Rm4mini() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'RM4MINI';
        return _this;
    }
    Rm4mini.prototype.checkSensors = function () {
        return this.send(0x24).then(function (payload) {
            return {
                temperature: (0, struct_1.default)('<bb').unpack(payload.subarray(0, 0x2))[0x0] + payload[0x1] / 100.0,
                humidity: payload[0x2] + payload[0x3] / 100.0,
            };
        });
    };
    Rm4mini.prototype.checkTemperature = function () {
        return this.checkSensors().then(function (_a) {
            var temperature = _a.temperature;
            return temperature;
        });
    };
    Rm4mini.prototype.checkHumidity = function () {
        return this.checkSensors().then(function (_a) {
            var humidity = _a.humidity;
            return humidity;
        });
    };
    return Rm4mini;
}(Rmminib));
exports.Rm4mini = Rm4mini;
var Rm4pro = /** @class */ (function (_super) {
    __extends(Rm4pro, _super);
    function Rm4pro() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rm4pro.prototype.sweepFrequency = function () {
        return this.send(0x19).then();
    };
    Rm4pro.prototype.cancelSweepFrequency = function () {
        return this.send(0x1e).then();
    };
    Rm4pro.prototype.checkFrequency = function () {
        return this.send(0x1a).then(function (response) { return response[0] == 1; });
    };
    Rm4pro.prototype.findRfPacket = function () {
        return this.send(0x1b).then();
    };
    Rm4pro.prototype.checkTemperature = function () {
        return this.checkSensors().then(function (_a) {
            var temperature = _a.temperature;
            return temperature;
        });
    };
    return Rm4pro;
}(Rm4mini));
exports.Rm4pro = Rm4pro;
