// node-broadlink/dist/sensor.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.A1 = void 0;
var device_1 = __importDefault(require("./device"));
var struct_1 = __importDefault(require("./struct"));
var A1 = /** @class */ (function (_super) {
    __extends(A1, _super);
    function A1() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'A1';
        return _this;
    }
    A1.prototype.checkSensors = function () {
        return this.checkSensorsRaw().then(function (state) {
            return __assign(__assign({}, state), { light: ['dark', 'dim', 'normal', 'bright'][state.light] || 'unknown', air_quality: ['excellent', 'good', 'normal', 'bad'][state.air_quality] || 'unknown', noise: ['quiet', 'normal', 'noisy'][state.noise] || 'unknown' });
        });
    };
    A1.prototype.checkSensorsRaw = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.sendPacket(Buffer.of(0x1))
                .then(function (response) {
                var payload = _this.decrypt(response);
                var data = payload.subarray(0x4);
                var temperatureBuffer = (0, struct_1.default)('<bb').unpack(data.subarray(0x0, 0x2));
                resolve({
                    temperature: temperatureBuffer[0x0] + temperatureBuffer[0x1] / 10.0,
                    humidity: data[0x2] + data[0x3] / 10.0,
                    light: data[0x4],
                    air_quality: data[0x6],
                    noise: data[0x8],
                });
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    return A1;
}(device_1.default));
exports.A1 = A1;
