// node-broadlink/dist/switch.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mp1 = exports.Bg1 = exports.Sp4b = exports.Sp4 = exports.Sp3s = exports.Sp3 = exports.Sp2s = exports.Sp2 = exports.Sp1 = void 0;
var device_1 = __importDefault(require("./device"));
var struct_1 = __importDefault(require("./struct"));
var Sp1 = /** @class */ (function (_super) {
    __extends(Sp1, _super);
    function Sp1() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'SP1';
        return _this;
    }
    Sp1.prototype.setPower = function (pwr) {
        var packet = Buffer.alloc(4);
        packet[0] = (pwr && 1) || 0;
        return this.sendPacket(packet).then();
    };
    return Sp1;
}(device_1.default));
exports.Sp1 = Sp1;
var Sp2 = /** @class */ (function (_super) {
    __extends(Sp2, _super);
    function Sp2() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'SP2';
        return _this;
    }
    Sp2.prototype.setPower = function (pwr) {
        var packet = Buffer.alloc(16);
        packet[0] = 2;
        packet[4] = (pwr && 1) || 0;
        return this.sendPacket(packet).then();
    };
    Sp2.prototype.checkPower = function () {
        var _this = this;
        var packet = Buffer.alloc(16);
        packet[0] = 1;
        return this.sendPacket(packet).then(function (response) {
            return !!_this.decrypt(response)[0x4];
        });
    };
    return Sp2;
}(device_1.default));
exports.Sp2 = Sp2;
var Sp2s = /** @class */ (function (_super) {
    __extends(Sp2s, _super);
    function Sp2s() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'SP2S';
        return _this;
    }
    Sp2s.prototype.getEnergy = function () {
        var packet = Buffer.alloc(16);
        packet[0] = 4;
        return this.sendPacketAndDecrypt(packet).then(function (payload) { return payload.subarray(0x4, 0x7).readInt8() / 1000; });
    };
    return Sp2s;
}(Sp2));
exports.Sp2s = Sp2s;
var Sp3 = /** @class */ (function (_super) {
    __extends(Sp3, _super);
    function Sp3() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'SP3';
        return _this;
    }
    Sp3.prototype.setPower = function (pwr) {
        var _this = this;
        return this.checkNightlight().then(function (bool) {
            var packet = Buffer.alloc(16);
            packet[0] = 2;
            packet[4] = (((bool && 1) || 0) << 1) | ((pwr && 1) || 0);
            return _this.sendPacket(packet).then();
        });
    };
    Sp3.prototype.setNightlight = function (ntlight) {
        var _this = this;
        return this.checkPower().then(function (bool) {
            var packet = Buffer.alloc(16);
            packet[0] = 2;
            packet[4] = (((ntlight && 1) || 0) << 1) | ((bool && 1) || 0);
            return _this.sendPacket(packet).then();
        });
    };
    Sp3.prototype.checkPower = function () {
        var _this = this;
        var packet = Buffer.alloc(16);
        packet[0] = 1;
        return this.sendPacket(packet).then(function (response) {
            var payload = _this.decrypt(response);
            return !!(payload[0x4] & 2);
        });
    };
    Sp3.prototype.checkNightlight = function () {
        var _this = this;
        var packet = Buffer.alloc(16);
        packet[0] = 1;
        return this.sendPacket(packet).then(function (response) {
            var payload = _this.decrypt(response);
            return !!(payload[0x4] & 1);
        });
    };
    return Sp3;
}(device_1.default));
exports.Sp3 = Sp3;
var Sp3s = /** @class */ (function (_super) {
    __extends(Sp3s, _super);
    function Sp3s() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'SP3S';
        return _this;
    }
    Sp3s.prototype.getEnergy = function () {
        return this.sendPacketAndDecrypt(Buffer.of(8, 0, 254, 1, 5, 1, 0, 0, 0, 45)).then(function (payload) { return payload.subarray(0x7a, 0x4).readInt8() / 100; });
    };
    return Sp3s;
}(Sp2));
exports.Sp3s = Sp3s;
var Sp4 = /** @class */ (function (_super) {
    __extends(Sp4, _super);
    function Sp4() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'SP4';
        return _this;
    }
    Sp4.prototype.setPower = function (pwr) {
        return this.setState({ pwr: pwr }).then();
    };
    Sp4.prototype.setNightlight = function (ntlight) {
        return this.setState({ ntlight: ntlight }).then();
    };
    Sp4.prototype.setState = function (state) {
        var _this = this;
        return this.sendPacket(this.encode(2, state)).then(function (payload) { return _this.decode(payload); });
    };
    Sp4.prototype.checkPower = function () {
        return this.getState().then(function (_a) {
            var pwr = _a.pwr;
            return pwr;
        });
    };
    Sp4.prototype.checkNightlight = function () {
        return this.getState().then(function (_a) {
            var ntlight = _a.ntlight;
            return ntlight;
        });
    };
    Sp4.prototype.getState = function () {
        var _this = this;
        return this.sendPacketAndDecrypt(this.encode(1, {})).then(function (payload) { return _this.decode(payload); });
    };
    Sp4.prototype.encode = function (flag, state) {
        var data = JSON.stringify(this.getValue(state, Number));
        var packet = Buffer.concat([
            (0, struct_1.default)('<HHHBBI').pack_into(Buffer.alloc(12), 0, 0xa5a5, 0x5a5a, 0x0000, flag, 0x0b, data.length),
            Buffer.from(data),
        ]);
        var checksum = packet.reduce(function (acc, v) { return acc + v; }, 0xbeaf) & 0xffff;
        packet[0x04] = checksum & 0xff;
        packet[0x05] = (checksum >> 8) & 0xff;
        return packet;
    };
    Sp4.prototype.getValue = function (state, Number) {
        return {
            pwr: state.pwr !== undefined ? Number(state.pwr) : undefined,
            ntlight: state.ntlight !== undefined ? Number(state.ntlight) : undefined,
            indicator: state.indicator !== undefined ? Number(state.indicator) : undefined,
            ntlbrightness: state.ntlbrightness,
            maxworktime: state.maxworktime,
            childlock: state.childlock !== undefined ? Number(state.childlock) : undefined,
        };
    };
    Sp4.prototype.decode = function (payload) {
        return this.getValue(JSON.parse(payload.subarray(0x0c, 0x0c + (0, struct_1.default)('<I').unpack_from(payload, 0x08)[0]).toString()), Boolean);
    };
    return Sp4;
}(device_1.default));
exports.Sp4 = Sp4;
var Sp4b = /** @class */ (function (_super) {
    __extends(Sp4b, _super);
    function Sp4b() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'SP4B';
        return _this;
    }
    Sp4b.prototype.encode = function (flag, state) {
        var data = JSON.stringify(this.getValue(state, Number));
        var packet = Buffer.concat([
            (0, struct_1.default)('<HHHHBBI').pack_into(Buffer.alloc(14), 0, 12 + data.length, 0xa5a5, 0x5a5a, 0x0000, flag, 0x0b, data.length),
            Buffer.from(data),
        ]);
        var checksum = packet.subarray(0x02).reduce(function (acc, b) { return acc + b; }, 0xbeaf) & 0xffff;
        packet[0x06] = checksum & 0xff;
        packet[0x07] = (checksum >> 8) & 0xff;
        return packet;
    };
    Sp4b.prototype.decode = function (payload) {
        var data = this.getValue(JSON.parse(payload.subarray(0x0e, 0x0e + (0, struct_1.default)('<I').unpack_from(payload, 0xa)[0]).toString()), Boolean);
        return __assign(__assign({}, data), { ntlbrightness: data.ntlbrightness !== undefined ? data.ntlbrightness / 1000 : undefined, maxworktime: data.maxworktime !== undefined ? data.maxworktime / 1000 : undefined });
    };
    return Sp4b;
}(Sp4));
exports.Sp4b = Sp4b;
var Bg1 = /** @class */ (function (_super) {
    __extends(Bg1, _super);
    function Bg1() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'BG1';
        return _this;
    }
    Bg1.prototype.getValue = function (state, Number) {
        return {
            pwr: state.pwr !== undefined ? Number(state.pwr) : undefined,
            pwr1: state.pwr1 !== undefined ? Number(state.pwr1) : undefined,
            pwr2: state.pwr2 !== undefined ? Number(state.pwr2) : undefined,
            maxworktime: state.maxworktime,
            maxworktime1: state.maxworktime1,
            maxworktime2: state.maxworktime2,
            idcbrightness: state.idcbrightness,
        };
    };
    Bg1.prototype.getState = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.sendPacket(_this.encode(1, {}))
                .then(function (response) {
                resolve(_this.decode(response));
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Bg1.prototype.setState = function (state) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.sendPacket(_this.encode(2, state))
                .then(function (response) {
                resolve(_this.decode(_this.decrypt(response)));
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Bg1.prototype.encode = function (flag, state) {
        var data = JSON.stringify(this.getValue(state, Number));
        var packet = Buffer.concat([
            Buffer.from((0, struct_1.default)('<HHHHBBI').pack_into(Buffer.alloc(14), 0, 12 + data.length, 0xa5a5, 0x5a5a, 0x0000, flag, 0x0b, data.length)),
            Buffer.from(data),
        ]);
        var checksum = packet.subarray(0x2).reduce(function (acc, b) { return acc + b; }, 0xbeaf) & 0xffff;
        packet[0x06] = checksum & 0xff;
        packet[0x07] = (checksum >> 8) & 0xff;
        return packet;
    };
    Bg1.prototype.decode = function (payload) {
        return this.getValue(JSON.parse(payload.subarray(0x0e, 0x0e + (0, struct_1.default)('<I').unpack_from(payload, 0x0a)[0]).toString()), Boolean);
    };
    return Bg1;
}(device_1.default));
exports.Bg1 = Bg1;
var Mp1 = /** @class */ (function (_super) {
    __extends(Mp1, _super);
    function Mp1() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.TYPE = 'MP1';
        return _this;
    }
    Mp1.prototype.setPowerMask = function (sidMask, state) {
        var packet = Buffer.alloc(16);
        packet[0x00] = 0x0d;
        packet[0x02] = 0xa5;
        packet[0x03] = 0xa5;
        packet[0x04] = 0x5a;
        packet[0x05] = 0x5a;
        packet[0x06] = 0xb2 + (state ? sidMask << 1 : sidMask);
        packet[0x07] = 0xc0;
        packet[0x08] = 0x02;
        packet[0x0a] = 0x03;
        packet[0x0d] = sidMask;
        packet[0x0e] = state ? sidMask : 0;
        return this.sendPacket(packet).then();
    };
    Mp1.prototype.setPower = function (sid, state) {
        return this.setPowerMask(0x01 << (sid - 1), Number(state)).then();
    };
    Mp1.prototype.checkPowerRaw = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var packet = Buffer.alloc(16);
            packet[0x00] = 0x0a;
            packet[0x02] = 0xa5;
            packet[0x03] = 0xa5;
            packet[0x04] = 0x5a;
            packet[0x05] = 0x5a;
            packet[0x06] = 0xae;
            packet[0x07] = 0xc0;
            packet[0x08] = 0x01;
            _this.sendPacket(packet)
                .then(function (response) {
                var payload = _this.decrypt(response);
                resolve(payload[0x0e]);
            })
                .catch(function (err) {
                reject(err);
            });
        });
    };
    Mp1.prototype.checkPower = function () {
        return this.checkPowerRaw().then(function (state) {
            return {
                s1: !!(state & 0x01),
                s2: !!(state & 0x02),
                s3: !!(state & 0x04),
                s4: !!(state & 0x08),
            };
        });
    };
    return Mp1;
}(device_1.default));
exports.Mp1 = Mp1;
